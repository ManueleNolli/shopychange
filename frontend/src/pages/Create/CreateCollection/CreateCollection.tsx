{
  /* eslint-disable */
}

import React, { useRef, useState } from 'react'

import {
  Box,
  Center,
  Flex,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Heading,
  Input,
  HStack,
  Spacer,
  Text,
  NumberInput,
  NumberInputField,
} from '@chakra-ui/react'
import { colors } from '../../../styles/Colors'

import { MdDelete } from 'react-icons/md'
import { TiPlus } from 'react-icons/ti'
import { BsFillCaretRightFill, BsFillCaretLeftFill } from 'react-icons/bs'

import UploadImage from '../../../components/UploadImage/UploadImage'
import LoadingButton from '../../../components/LoadingButton/LoadingButton'
import ShadowButton from '../../../components/ShadowButton/ShadowButton'
import { Formik, Form, Field, FieldArray } from 'formik'

import { Attribute, NewNFT } from '../../../types/components/NewNFT'

import { validatePresence } from '../../../utils/ValidatePresence/validatePresence'
import { validateImage } from '../../../utils/ValidateImage/validateImage'
import { Royalty } from '../../../types/components/Royalty'
import { validateAddress } from '../../../utils/ValidateAddress/ValidateAddress'
import {
  checkIfRoyaltyReceiverIsInvalid,
  errorMessageRoyaltyReceiver,
  checkIfRoyaltyValuesIsInvalid,
  checkIfLastRoyalty,
} from '../../../utils/FormikRoyalty/FormikRoyalty'
import useCreateCollection from './useCreateCollection'

export interface FormValues {
  name: string
  symbol: string
  nfts: NewNFT[]
  royalties: Royalty[]
}

export default function CreateCollection() {
  const { isLoading, handleSubmit } = useCreateCollection()

  const [currentNftIndex, setCurrentNftIndex] = useState<number>(0)

  const containerRef = useRef<HTMLDivElement>(null)
  const addNFTref = useRef<HTMLButtonElement>(null)

  const scrollLeft = () => {
    if (containerRef.current && currentNftIndex > 0) {
      setCurrentNftIndex(currentNftIndex - 1)
      containerRef.current.scrollLeft -= window.innerWidth * 0.5 // Adjust the scroll distance as needed
    }
  }

  const scrollRight = () => {
    if (containerRef.current) {
      setCurrentNftIndex(currentNftIndex + 1)
      containerRef.current.scrollLeft += window.innerWidth * 0.5 // Adjust the scroll distance as needed
    }
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const handleAddNFT = (push: any, scrollRight: any) => {
    push({
      name: '',
      description: '',
      image: new File([], ''),
      attributes: [
        {
          trait_type: '',
          value: '',
        },
      ],
    })
    // HACK: Wait for the new NFT to be added to the DOM before scrolling
    setTimeout(() => {
      scrollRight()
    }, 0)
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const handleRemoveNFT = (remove: any, index: number) => {
    setCurrentNftIndex(currentNftIndex - 1)
    remove(index)
  }

  return (
    <Box mb="4vh">
      <Center>
        <Heading>New Collection</Heading>
      </Center>

      <Center mt="2vh">
        <Box w={{ base: '90vw', md: '50vw' }}>
          <Formik
            initialValues={{
              name: '',
              symbol: '',
              royalties: [
                {
                  receiver: '',
                  share: NaN,
                },
              ],
              nfts: [
                {
                  name: '',
                  description: '',
                  image: new File([], ''),
                  attributes: [
                    {
                      trait_type: '',
                      value: '',
                    },
                  ],
                },
              ],
            }}
            onSubmit={handleSubmit}
          >
            {/* eslint-disable-next-line */}
            {({ values }: any) => (
              <Form>
                <Field
                  name="name"
                  validate={(value: string) => validatePresence('Name', value)}
                >
                  {/* eslint-disable-next-line */}
                  {({ field, form }: any) => (
                    <FormControl
                      isRequired
                      isInvalid={form.errors.name && form.touched.name}
                    >
                      <FormLabel htmlFor="name">Name</FormLabel>
                      <Input {...field} id="name" placeholder="name" />
                      <FormErrorMessage>{form.errors.name}</FormErrorMessage>
                    </FormControl>
                  )}
                </Field>

                <Field
                  name="symbol"
                  validate={(value: string) =>
                    validatePresence('symbol', value)
                  }
                >
                  {/* eslint-disable-next-line */}
                  {({ field, form }: any) => (
                    <FormControl
                      isRequired
                      mt="2"
                      isInvalid={form.errors.symbol && form.touched.symbol}
                    >
                      <FormLabel htmlFor="symbol">Symbol</FormLabel>
                      <Input {...field} id="symbol" placeholder="symbol" />
                      <FormErrorMessage>{form.errors.symbol}</FormErrorMessage>
                    </FormControl>
                  )}
                </Field>

                <FieldArray name={'royalties'} validateOnChange={true}>
                  {({
                    push: pushRoyalty,
                    remove: removeRoyalty,
                  }: /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
                  any) => (
                    <Box mt="2">
                      <FormLabel>Royalties</FormLabel>
                      {values.royalties.map(
                        (value: Royalty, indexAttribute: number) => (
                          <Flex key={`royalties[${indexAttribute}]`} mt="2">
                            <Box w="70%">
                              <Field
                                key={`royalties[${indexAttribute}].receiver`}
                                name={`royalties[${indexAttribute}].receiver`}
                                validate={(value: string) =>
                                  validateAddress(value)
                                }
                              >
                                {/* eslint-disable-next-line */}
                                {({ field, form }: any) => (
                                  <FormControl
                                    isInvalid={checkIfRoyaltyReceiverIsInvalid(
                                      form,
                                      indexAttribute
                                    )}
                                  >
                                    <Input
                                      {...field}
                                      id={`royalties[${indexAttribute}].receiver`}
                                      placeholder="Receiver"
                                    />
                                    <FormErrorMessage>
                                      {errorMessageRoyaltyReceiver(
                                        form,
                                        indexAttribute
                                      )}
                                    </FormErrorMessage>
                                  </FormControl>
                                )}
                              </Field>
                            </Box>

                            <Spacer />
                            <Box w="20%">
                              <Field
                                key={`royalties[${indexAttribute}].share`}
                                name={`royalties[${indexAttribute}].share`}
                              >
                                {/* eslint-disable-next-line */}
                                {({ field, form }: any) => (
                                  <FormControl
                                    isInvalid={checkIfRoyaltyValuesIsInvalid(
                                      form
                                    )}
                                  >
                                    <NumberInput
                                      min={0}
                                      max={100}
                                      precision={2}
                                      step={0.01}
                                    >
                                      <NumberInputField
                                        {...field}
                                        id={`royalties[${indexAttribute}].share`}
                                        placeholder="Share (%)"
                                      />
                                    </NumberInput>
                                    {checkIfLastRoyalty(
                                      form,
                                      indexAttribute
                                    ) ? (
                                      <FormErrorMessage>
                                        Percentage exceed 100%
                                      </FormErrorMessage>
                                    ) : null}
                                  </FormControl>
                                )}
                              </Field>
                            </Box>

                            <Center w="5%">
                              <Box maxW="36px">
                                <MdDelete
                                  aria-label="remove-royalty"
                                  onClick={() => removeRoyalty(indexAttribute)}
                                  fill="#FF4500"
                                  size={'80%'}
                                />
                              </Box>
                            </Center>
                          </Flex>
                        )
                      )}
                      <Box mt="2" h="2.5vh" w="2.5vh">
                        <Box maxW="24px">
                          <TiPlus
                            aria-label="add-royalty"
                            fill="#228B22"
                            size={'100%'}
                            onClick={() =>
                              pushRoyalty({
                                receiver: '',
                                share: 0,
                              })
                            }
                          />
                        </Box>
                      </Box>
                    </Box>
                  )}
                </FieldArray>

                <Flex
                  my="2vh"
                  py="2vh"
                  border="1px"
                  borderColor="gray.200"
                  borderRadius={10}
                  w={{ base: '100vw', md: '60vw' }}
                  ml={{ base: '-5vw', md: '-5vw' }}
                >
                  <Center>
                    {values.nfts.length > 0 &&
                      (currentNftIndex > 0 ? (
                        <BsFillCaretLeftFill
                          aria-label="scroll-left"
                          style={{
                            width: '5vw',
                          }}
                          size={'2.5vw'}
                          onClick={scrollLeft}
                        />
                      ) : (
                        <Box w="5vw" />
                      ))}
                  </Center>

                  <Box
                    w={{ base: '90vw', md: '50vw' }}
                    hidden={values.nfts.length > 0 ? false : true}
                  >
                    <HStack
                      ref={containerRef}
                      spacing={0}
                      sx={{
                        '::-webkit-scrollbar': {
                          display: 'none',
                        },
                        scrollBehavior: 'smooth',
                        touchAction: 'none',
                        overflowY: 'hidden',
                        overflowX: 'scroll',
                        msOverflowStyle: 'none', // IE and Edge
                        scrollbarWidth: 'none', // Firefox
                      }}
                    >
                      <FieldArray name="nfts">
                        {/* eslint-disable-next-line */}
                        {({ push, remove }: any) => (
                          <>
                            {values.nfts.map((nft: NewNFT, index: number) => (
                              <Box
                                key={index}
                                minW={{ base: '90vw', md: '50vw' }}
                              >
                                <Center>
                                  <Heading size="md">Nft #{index}</Heading>
                                  <Box maxW="36px">
                                    <MdDelete
                                      aria-label="remove-nft"
                                      style={{
                                        marginLeft: '1vw',
                                      }}
                                      size={'100%'}
                                      fill="#FF4500"
                                      onClick={() =>
                                        handleRemoveNFT(remove, index)
                                      }
                                    />
                                  </Box>
                                </Center>
                                <Center>
                                  <Text size="xs" as="i" my="1vh">
                                    Note: Using the charachter{' '}
                                    <Text as="b">{'{id}'}</Text> in the name or
                                    description will be replaced by the NFT id
                                  </Text>
                                </Center>

                                <Field
                                  name={`nfts.${index}.image`}
                                  validate={(value: File) =>
                                    validateImage(value)
                                  }
                                >
                                  {/* eslint-disable-next-line */}
                                  {({ field, form }: any) => (
                                    <UploadImage
                                      currentImage={
                                        values.nfts[index].image &&
                                        values.nfts[index].image.size > 0
                                          ? values.nfts[index].image
                                          : undefined
                                      }
                                      /* eslint-disable-next-line */
                                      onUpload={(event: any) => {
                                        form.setFieldValue(
                                          `nfts.${index}.image`,
                                          event.currentTarget.files[0]
                                        )
                                      }}
                                      cancelUpload={() => {
                                        form.setFieldValue(
                                          `nfts.${index}.image`,
                                          new File([], '')
                                        )
                                      }}
                                    />
                                  )}
                                </Field>

                                <Field
                                  name={`nfts.${index}.name`}
                                  mt="2vh"
                                  validate={(value: string) =>
                                    validatePresence(`Name NFT ${index}`, value)
                                  }
                                >
                                  {/* eslint-disable-next-line */}
                                  {({ field, form }: any) => (
                                    <FormControl isRequired>
                                      <FormLabel htmlFor="name">Name</FormLabel>
                                      <Input
                                        {...field}
                                        id={`nfts.${index}.name`}
                                        placeholder="name"
                                      />
                                    </FormControl>
                                  )}
                                </Field>

                                <Field
                                  name={`nfts.${index}.description`}
                                  validate={(value: string) =>
                                    validatePresence(
                                      `Description NFT ${index}`,
                                      value
                                    )
                                  }
                                >
                                  {/* eslint-disable-next-line */}
                                  {({ field, form }: any) => (
                                    <FormControl isRequired>
                                      <FormLabel htmlFor="description">
                                        Description
                                      </FormLabel>
                                      <Input
                                        {...field}
                                        id={`nfts.${index}.description`}
                                        placeholder="description"
                                      />
                                    </FormControl>
                                  )}
                                </Field>

                                <FieldArray name={`nfts[${index}].attributes`}>
                                  {({
                                    push: pushAttribute,
                                    remove: removeAttribute,
                                  }: /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
                                  any) => (
                                    <Box mt="2">
                                      <FormLabel>Attributes</FormLabel>
                                      {values.nfts[index].attributes.map(
                                        (
                                          value: Attribute,
                                          indexAttribute: number
                                        ) => (
                                          <Flex
                                            key={`nfts[${index}].attributes[${indexAttribute}]`}
                                            mt="2"
                                          >
                                            <Box w="45%">
                                              <Field
                                                key={`nfts[${index}].attributes[${indexAttribute}].trait_type`}
                                                name={`nfts[${index}].attributes[${indexAttribute}].trait_type`}
                                              >
                                                {/* eslint-disable-next-line */}
                                                {({ field, form }: any) => (
                                                  <FormControl>
                                                    <Input
                                                      {...field}
                                                      id={`nfts[${index}].attributes[${indexAttribute}].trait_type`}
                                                      placeholder="trait type"
                                                    />
                                                  </FormControl>
                                                )}
                                              </Field>
                                            </Box>

                                            <Spacer />
                                            <Box w="45%">
                                              <Field
                                                key={`nfts[${index}].attributes[${indexAttribute}].value`}
                                                name={`nfts[${index}].attributes[${indexAttribute}].value`}
                                              >
                                                {/* eslint-disable-next-line */}
                                                {({ field, form }: any) => (
                                                  <FormControl>
                                                    <Input
                                                      {...field}
                                                      id={`nfts[${index}]-attributes[${indexAttribute}].value`}
                                                      placeholder="value"
                                                    />
                                                  </FormControl>
                                                )}
                                              </Field>
                                            </Box>

                                            <Center w="5%">
                                              <Box maxW="36px">
                                                <MdDelete
                                                  aria-label="remove-attribute"
                                                  onClick={() =>
                                                    removeAttribute(
                                                      indexAttribute
                                                    )
                                                  }
                                                  fill="#FF4500"
                                                  size={'80%'}
                                                />
                                              </Box>
                                            </Center>
                                          </Flex>
                                        )
                                      )}
                                      <Box mt="2" h="2.5vh" w="2.5vh">
                                        <Box maxW="24px">
                                          <TiPlus
                                            aria-label="add-attribute"
                                            fill="#228B22"
                                            size={'100%'}
                                            onClick={() =>
                                              pushAttribute({
                                                trait_type: '',
                                                value: '',
                                              })
                                            }
                                          />
                                        </Box>
                                      </Box>
                                    </Box>
                                  )}
                                </FieldArray>
                              </Box>
                            ))}
                            <button
                              ref={addNFTref}
                              type="button"
                              onClick={() => handleAddNFT(push, scrollRight)}
                            />
                          </>
                        )}
                      </FieldArray>
                    </HStack>
                  </Box>
                  <Center>
                    {values.nfts.length > 0 &&
                      (values.nfts.length === currentNftIndex + 1 ? (
                        <Box maxW="24px">
                          <TiPlus
                            aria-label="add-nft"
                            fill="#228B22"
                            style={{
                              width: '5vw',
                            }}
                            size={'2.5vw'}
                            onClick={() => {
                              if (addNFTref.current) addNFTref.current.click()
                            }}
                          />
                        </Box>
                      ) : (
                        <BsFillCaretRightFill
                          style={{
                            width: '5vw',
                          }}
                          size={'2.5vw'}
                          onClick={scrollRight}
                        />
                      ))}
                  </Center>

                  {values.nfts.length === 0 && (
                    <Center flex="1">
                      <Box maxW="24px">
                        <TiPlus
                          aria-label="add-nft-when-empty"
                          fill="#228B22"
                          style={{
                            width: '5vw',
                          }}
                          size={'2.5vw'}
                          onClick={() => {
                            if (addNFTref.current) addNFTref.current.click()
                          }}
                        />
                      </Box>
                    </Center>
                  )}
                </Flex>

                <Center>
                  <LoadingButton
                    isLoading={isLoading}
                    loadingButtonProps={{
                      size: 'lg',
                      width: '50%',
                    }}
                  >
                    <ShadowButton
                      label={'Create Collection'}
                      bgColor={`${colors.sellChakraColor}.400`}
                      textColor={'white'}
                      shadowColor={colors.sellShadowColor}
                      hoverColor={`${colors.sellChakraColor}.500`}
                      focusColor={`${colors.sellChakraColor}.500`}
                      type="submit"
                      buttonProps={{
                        size: 'lg',
                        width: '50%',
                      }}
                    />
                  </LoadingButton>
                </Center>
              </Form>
            )}
          </Formik>
        </Box>
      </Center>
    </Box>
  )
}
