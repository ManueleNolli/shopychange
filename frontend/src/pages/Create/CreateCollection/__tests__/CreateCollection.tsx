import React from 'react'
import { render, screen, fireEvent, act } from '@testing-library/react'
import useCreateCollection from '../useCreateCollection'
import CreateCollection from '../CreateCollection'
import { Logo } from '../../../../assets/AssetsManager'
import { validateImage } from '../../../../utils/ValidateImage/validateImage'

const mockCreateObjectURL = jest.fn()
URL.createObjectURL = mockCreateObjectURL

jest.mock('../../../../utils/ValidateImage/validateImage', () => ({
  validateImage: jest.fn(),
}))

jest.mock('../useCreateCollection', () => jest.fn())

describe('CreateCollection', () => {
  test('renders correctly', () => {
    ;(useCreateCollection as jest.Mock).mockReturnValue({
      isLoading: false,
      handleSubmit: jest.fn(),
    })

    const tree = render(<CreateCollection />)

    expect(tree.container).toMatchSnapshot()
  })

  test('should add new nft', async () => {
    ;(useCreateCollection as jest.Mock).mockReturnValue({
      isLoading: false,
      handleSubmit: jest.fn(),
    })
    render(<CreateCollection />)
    const description = screen.getByPlaceholderText('description')
    await act(async () => {
      fireEvent.change(description, { target: { value: 'test' } })
    })
    const addNFTButton = screen.getByLabelText('add-nft')

    await act(async () => {
      fireEvent.click(addNFTButton)
    })

    const description2 = screen.getAllByPlaceholderText('description')

    expect(description2).toHaveLength(2)

    const valueFirstNFTDescription = description2[0].getAttribute('value')
    expect(valueFirstNFTDescription).toBe('test')

    const valueSecondNFTDescription = description2[1].getAttribute('value')
    expect(valueSecondNFTDescription).toBe('')
  })

  test('should add new nft when empty', async () => {
    ;(useCreateCollection as jest.Mock).mockReturnValue({
      isLoading: false,
      handleSubmit: jest.fn(),
    })
    render(<CreateCollection />)
    const removeNFTButton = screen.getByLabelText('remove-nft')

    await act(async () => {
      fireEvent.click(removeNFTButton)
    })

    const description = screen.queryByPlaceholderText('description')
    expect(description).toBeNull()

    const addNFTButton = screen.getByLabelText('add-nft-when-empty')

    await act(async () => {
      fireEvent.click(addNFTButton)
    })

    const description2 = screen.getAllByPlaceholderText('description')
    expect(description2).toHaveLength(1)
  })

  test('should remove nft', async () => {
    ;(useCreateCollection as jest.Mock).mockReturnValue({
      isLoading: false,
      handleSubmit: jest.fn(),
    })
    render(<CreateCollection />)
    const description = screen.getByPlaceholderText('description')

    await act(async () => {
      fireEvent.change(description, { target: { value: 'test' } })
    })
    const removeNFTButton = screen.getByLabelText('remove-nft')

    await act(async () => {
      fireEvent.click(removeNFTButton)
    })

    const description2 = screen.queryByPlaceholderText('description')
    expect(description2).toBeNull()
  })

  test('should scroll to left to previous nft', async () => {
    ;(useCreateCollection as jest.Mock).mockReturnValue({
      isLoading: false,
      handleSubmit: jest.fn(),
    })
    render(<CreateCollection />)

    const addNFTButton = screen.getByLabelText('add-nft')

    await act(async () => {
      fireEvent.click(addNFTButton)
    })

    const scrollLeftButton = screen.getByLabelText('scroll-left')

    await act(async () => {
      fireEvent.click(scrollLeftButton)
    })
  })

  test('should handle add and remove royalties', async () => {
    ;(useCreateCollection as jest.Mock).mockReturnValue({
      isLoading: false,
      handleSubmit: jest.fn(),
    })

    render(<CreateCollection />)

    const addAttributeButton = screen.getByLabelText('add-royalty')

    await act(async () => {
      fireEvent.click(addAttributeButton)
    })

    const placeholder = screen.getAllByPlaceholderText('Share (%)').length

    expect(placeholder).toBe(2)

    const removeRoyaltyButton = screen.getAllByLabelText('remove-royalty')

    await act(async () => {
      fireEvent.click(removeRoyaltyButton[0])
    })

    const placeholderAfterRemove =
      screen.getAllByPlaceholderText('Share (%)').length

    expect(placeholderAfterRemove).toBe(1)
  })
  test('should handle add and remove attributes', async () => {
    ;(useCreateCollection as jest.Mock).mockReturnValue({
      isLoading: false,
      handleSubmit: jest.fn(),
    })

    render(<CreateCollection />)

    const addAttributeButton = screen.getByLabelText('add-attribute')

    await act(async () => {
      fireEvent.click(addAttributeButton)
    })

    const traitTypeCount = screen.getAllByPlaceholderText('trait type').length
    const traitValueCount = screen.getAllByPlaceholderText('value').length

    expect(traitTypeCount).toBe(2)
    expect(traitValueCount).toBe(2)

    const removeAttributeButton = screen.getAllByLabelText('remove-attribute')

    await act(async () => {
      fireEvent.click(removeAttributeButton[0])
    })

    const traitTypeCountAfterRemove =
      screen.getAllByPlaceholderText('trait type').length
    const traitValueCountAfterRemove =
      screen.getAllByPlaceholderText('value').length

    expect(traitTypeCountAfterRemove).toBe(1)
    expect(traitValueCountAfterRemove).toBe(1)
  })

  test('should validate image', async () => {
    ;(validateImage as jest.Mock).mockResolvedValueOnce(true)
    ;(useCreateCollection as jest.Mock).mockReturnValue({
      isLoading: false,
      handleSubmit: jest.fn(),
    })

    render(<CreateCollection />)

    const imageInput = screen.getByTestId('image-input')

    // use logo as image with mocked size
    Object.defineProperty(imageInput, 'files', {
      value: [new File([Logo], 'filename', { type: 'image/png' })],
    })

    await act(async () => {
      fireEvent.change(imageInput)
    })

    //should have called validateImage
    expect(validateImage).toHaveBeenCalled()

    const deleteButton = screen.getByTestId('delete-button')

    await act(async () => {
      fireEvent.click(deleteButton)
    })

    //should have called validateImage
    expect(validateImage).toHaveBeenCalled()
  })
})
