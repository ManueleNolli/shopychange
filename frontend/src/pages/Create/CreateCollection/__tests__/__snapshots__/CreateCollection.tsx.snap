// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCollection renders correctly 1`] = `
<div>
  <div
    class="css-q5vkps"
  >
    <div
      class="css-gmuwbf"
    >
      <h2
        class="chakra-heading css-0"
      >
        New Collection
      </h2>
    </div>
    <div
      class="css-mr3wqg"
    >
      <div
        class="css-f4ygcn"
      >
        <form
          action="#"
        >
          <div
            class="chakra-form-control css-0"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="name"
              id="field-:r0:-label"
            >
              Name
              <span
                aria-hidden="true"
                class="chakra-form__required-indicator css-0"
                role="presentation"
              >
                *
              </span>
            </label>
            <input
              aria-required="true"
              class="chakra-input css-0"
              id="name"
              name="name"
              placeholder="name"
              required=""
              value=""
            />
          </div>
          <div
            class="chakra-form-control css-g89h0y"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              for="symbol"
              id="field-:r1:-label"
            >
              Symbol
              <span
                aria-hidden="true"
                class="chakra-form__required-indicator css-0"
                role="presentation"
              >
                *
              </span>
            </label>
            <input
              aria-required="true"
              class="chakra-input css-0"
              id="symbol"
              name="symbol"
              placeholder="symbol"
              required=""
              value=""
            />
          </div>
          <div
            class="css-g89h0y"
          >
            <label
              class="chakra-form__label css-uvho5g"
            >
              Royalties
            </label>
            <div
              class="css-bj7zsy"
            >
              <div
                class="css-6gs9hh"
              >
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <input
                    class="chakra-input css-0"
                    id="royalties[0].receiver"
                    name="royalties[0].receiver"
                    placeholder="Receiver"
                    value=""
                  />
                </div>
              </div>
              <div
                class="css-17xejub"
              />
              <div
                class="css-ufscxh"
              >
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <div
                    class="chakra-numberinput css-e69dqy"
                  >
                    <input
                      aria-readonly="false"
                      aria-required="false"
                      aria-valuemax="100"
                      aria-valuemin="0"
                      autocomplete="off"
                      autocorrect="off"
                      class="chakra-numberinput__field css-8atqhb"
                      id="royalties[0].share"
                      inputmode="decimal"
                      name="royalties[0].share"
                      pattern="[0-9]*(.[0-9]+)?"
                      placeholder="Share (%)"
                      role="spinbutton"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                class="css-8tf3yw"
              >
                <div
                  class="css-cse5we"
                >
                  <svg
                    aria-label="remove-royalty"
                    fill="#FF4500"
                    height="80%"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="80%"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="css-98x68k"
            >
              <div
                class="css-913hcs"
              >
                <svg
                  aria-label="add-royalty"
                  baseProfile="tiny"
                  fill="#228B22"
                  height="100%"
                  stroke="currentColor"
                  stroke-width="0"
                  version="1.2"
                  viewBox="0 0 24 24"
                  width="100%"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M18 10h-4v-4c0-1.104-.896-2-2-2s-2 .896-2 2l.071 4h-4.071c-1.104 0-2 .896-2 2s.896 2 2 2l4.071-.071-.071 4.071c0 1.104.896 2 2 2s2-.896 2-2v-4.071l4 .071c1.104 0 2-.896 2-2s-.896-2-2-2z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="css-19twi6u"
          >
            <div
              class="css-gmuwbf"
            >
              <div
                class="css-1l9qw8v"
              />
            </div>
            <div
              class="css-f4ygcn"
            >
              <div
                class="chakra-stack css-17ftfu3"
              >
                <div
                  class="css-yk4jyz"
                >
                  <div
                    class="css-gmuwbf"
                  >
                    <h2
                      class="chakra-heading css-0"
                    >
                      Nft #
                      0
                    </h2>
                    <div
                      class="css-cse5we"
                    >
                      <svg
                        aria-label="remove-nft"
                        fill="#FF4500"
                        height="100%"
                        stroke="currentColor"
                        stroke-width="0"
                        style="margin-left: 1vw;"
                        viewBox="0 0 24 24"
                        width="100%"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M0 0h24v24H0z"
                          fill="none"
                        />
                        <path
                          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    class="css-gmuwbf"
                  >
                    <i
                      class="chakra-text css-1fa2p9l"
                    >
                      Note: Using the charachter
                       
                      <b
                        class="chakra-text css-0"
                      >
                        {id}
                      </b>
                       in the name or description will be replaced by the NFT id
                    </i>
                  </div>
                  <div
                    class="chakra-form-control css-0"
                    role="group"
                  >
                    <label
                      class="chakra-form__label css-uvho5g"
                      for="image"
                      id="field-:r4:-label"
                    >
                      Image
                      <span
                        aria-hidden="true"
                        class="chakra-form__required-indicator css-0"
                        role="presentation"
                      >
                        *
                      </span>
                    </label>
                    <div
                      class="chakra-aspect-ratio css-oj014t"
                    >
                      <div
                        class="chakra-input__group css-bx0blc"
                        data-group="true"
                      >
                        <input
                          accept="image/*"
                          data-testid="image-input"
                          id="image"
                          placeholder="image"
                          style="display: none;"
                          type="file"
                        />
                        <img
                          class="chakra-image css-glhh6s"
                          data-testid="image-placeholder"
                          src="placeholder.png"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="chakra-form-control css-0"
                    role="group"
                  >
                    <label
                      class="chakra-form__label css-uvho5g"
                      for="name"
                      id="field-:r5:-label"
                    >
                      Name
                      <span
                        aria-hidden="true"
                        class="chakra-form__required-indicator css-0"
                        role="presentation"
                      >
                        *
                      </span>
                    </label>
                    <input
                      aria-required="true"
                      class="chakra-input css-0"
                      id="nfts.0.name"
                      name="nfts.0.name"
                      placeholder="name"
                      required=""
                      value=""
                    />
                  </div>
                  <div
                    class="chakra-form-control css-0"
                    role="group"
                  >
                    <label
                      class="chakra-form__label css-uvho5g"
                      for="description"
                      id="field-:r6:-label"
                    >
                      Description
                      <span
                        aria-hidden="true"
                        class="chakra-form__required-indicator css-0"
                        role="presentation"
                      >
                        *
                      </span>
                    </label>
                    <input
                      aria-required="true"
                      class="chakra-input css-0"
                      id="nfts.0.description"
                      name="nfts.0.description"
                      placeholder="description"
                      required=""
                      value=""
                    />
                  </div>
                  <div
                    class="css-g89h0y"
                  >
                    <label
                      class="chakra-form__label css-uvho5g"
                    >
                      Attributes
                    </label>
                    <div
                      class="css-bj7zsy"
                    >
                      <div
                        class="css-1ttywp7"
                      >
                        <div
                          class="chakra-form-control css-0"
                          role="group"
                        >
                          <input
                            class="chakra-input css-0"
                            id="nfts[0].attributes[0].trait_type"
                            name="nfts[0].attributes[0].trait_type"
                            placeholder="trait type"
                            value=""
                          />
                        </div>
                      </div>
                      <div
                        class="css-17xejub"
                      />
                      <div
                        class="css-1ttywp7"
                      >
                        <div
                          class="chakra-form-control css-0"
                          role="group"
                        >
                          <input
                            class="chakra-input css-0"
                            id="nfts[0]-attributes[0].value"
                            name="nfts[0].attributes[0].value"
                            placeholder="value"
                            value=""
                          />
                        </div>
                      </div>
                      <div
                        class="css-8tf3yw"
                      >
                        <div
                          class="css-cse5we"
                        >
                          <svg
                            aria-label="remove-attribute"
                            fill="#FF4500"
                            height="80%"
                            stroke="currentColor"
                            stroke-width="0"
                            viewBox="0 0 24 24"
                            width="80%"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M0 0h24v24H0z"
                              fill="none"
                            />
                            <path
                              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                    <div
                      class="css-98x68k"
                    >
                      <div
                        class="css-913hcs"
                      >
                        <svg
                          aria-label="add-attribute"
                          baseProfile="tiny"
                          fill="#228B22"
                          height="100%"
                          stroke="currentColor"
                          stroke-width="0"
                          version="1.2"
                          viewBox="0 0 24 24"
                          width="100%"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M18 10h-4v-4c0-1.104-.896-2-2-2s-2 .896-2 2l.071 4h-4.071c-1.104 0-2 .896-2 2s.896 2 2 2l4.071-.071-.071 4.071c0 1.104.896 2 2 2s2-.896 2-2v-4.071l4 .071c1.104 0 2-.896 2-2s-.896-2-2-2z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
                <button
                  type="button"
                />
              </div>
            </div>
            <div
              class="css-gmuwbf"
            >
              <div
                class="css-913hcs"
              >
                <svg
                  aria-label="add-nft"
                  baseProfile="tiny"
                  fill="#228B22"
                  height="2.5vw"
                  stroke="currentColor"
                  stroke-width="0"
                  style="width: 5vw;"
                  version="1.2"
                  viewBox="0 0 24 24"
                  width="2.5vw"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M18 10h-4v-4c0-1.104-.896-2-2-2s-2 .896-2 2l.071 4h-4.071c-1.104 0-2 .896-2 2s.896 2 2 2l4.071-.071-.071 4.071c0 1.104.896 2 2 2s2-.896 2-2v-4.071l4 .071c1.104 0 2-.896 2-2s-.896-2-2-2z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="css-gmuwbf"
          >
            <button
              class="chakra-button css-187rakm"
              data-testid="onClick-button"
              style="white-space: initial;"
              type="submit"
            >
              Create Collection
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
